src/safemode_ui.h

```diff
@@ -0,0 +1,124 @@
+#ifndef SEFEMODE_UI_H
+#define SEFEMODE_UI_H
+
+#include <unordered_map>
+#include <string>
+#include <vector>
+#include <locale>
+#include <algorithm>
+#include "json.h"
+#include "enums.h"
+#include "creature.h"
+
+class safemode : public JsonSerializer, public JsonDeserializer
+{
+    private:
+        enum Tabs : int {
+            GLOBAL_TAB,
+            CHARACTER_TAB,
+            MAX_TAB
+        };
+
+        class rules_class
+        {
+            public:
+                std::string rule;
+                bool active;
+                bool exclude;
+                Creature::Attitude attitude;
+                int proximity;
+
+                rules_class() {
+                    this->rule = "";
+                    this->active = false;
+                    this->exclude = false;
+                    this->attitude = Creature::A_HOSTILE;
+                    this->proximity = 0;
+                }
+
+                rules_class( std::string rule_in, bool active_in, bool exclude_in,
+                             Creature::Attitude attitude_in, int proximity_in ) {
+                    this->rule = rule_in;
+                    this->active = active_in;
+                    this->exclude = exclude_in;
+                    this->attitude = attitude_in;
+                    this->proximity = proximity_in;
+                }
+
+                ~rules_class() {};
```

Comment:
If the destructor is empty (and non-virtual), don't declare it. The implicit compiler generated one is just as good.